### PR TITLE
fix: track uToken supply manually

### LINKED ISSUE #505: x/leverage: fix total uToken supply + ibc interaction
Found in trail of bits audit.

> Umee specification states that the Token:uToken exchange rate can only be affected by interest accruing (and not by Lend, Withdraw, Borrow, Repay, and Liquidate transactions). However, there are two methods to break this invariant:
> 
> - Burning (or minting) tokens via the Inter-Blockchain Communication (IBC) transfers. The ibc-go library accesses x/bank module’s keeper interface, which changes the total supply – as can be seen in figure 21.2. Please note that the issue is marked in a comment, as shown in figure 22.2.
> - Sending tokens directly to the module with x/bank messages.

The second case is effectively an attacker "donating" funds to uToken holders with no negative effect. The first, on the other hand, is a vulnerability that must be addressed (seen in our existing TODO comment).

```
func (k Keeper) TotalUTokenSupply(ctx sdk.Context, uTokenDenom string) sdk.Coin {
  if k.IsAcceptedUToken(ctx, uTokenDenom) {
      return k.bankKeeper.GetSupply(ctx, uTokenDenom)
      // TODO - Question: Does bank module still track balances sent (locked) via IBC?
      // If it doesn't then the balance returned here would decrease when the tokens
      // are sent off, which is not what we want. In that case, the keeper should keep
      // an sdk.Int total supply for each uToken type.
  }
  return sdk.NewCoin(uTokenDenom, sdk.ZeroInt())
}
```

We should implement the solution described in the comment to avoid `ibc + bank` interactions

### PR DESCRIPTION
**Pull request by toteki**
## Description

Fixes issue found in TrailOfBits audit

Closes #505 

----

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [x] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [x] added appropriate labels to the PR
- [ ] added `!` to the type prefix if API or client breaking change
- [x] targeted the correct branch (see [PR Targeting](https://github.com/umee-network/umee/blob/main/CONTRIBUTING.md#pr-targeting))
- [x] provided a link to the relevant issue or specification
- [ ] added a changelog entry to `CHANGELOG.md`
- [x] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [x] updated the relevant documentation or specification
- [x] reviewed "Files changed" and left comments if necessary
- [x] confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed `!` in the type prefix if API or client breaking change
- [ ] confirmed all author checklist items have been addressed
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

### CODE REVIEW

File: proto/umee/leverage/v1beta1/genesis.proto [hunk 1]
```diff
@@ -16,9 +16,11 @@ message GenesisState {
   repeated Collateral               collateral          = 5 [(gogoproto.nullable) = false];
   repeated cosmos.base.v1beta1.Coin reserves            = 6
       [(gogoproto.nullable) = false, (gogoproto.castrepeated) = "github.com/cosmos/cosmos-sdk/types.Coins"];
-  int64                           last_interest_time = 7;
-  repeated BadDebt                bad_debts          = 8 [(gogoproto.nullable) = false];
-  repeated InterestScalar         interest_scalars   = 9 [(gogoproto.nullable) = false];
+  int64                             last_interest_time = 7;
+  repeated BadDebt                  bad_debts          = 8 [(gogoproto.nullable) = false];
+  repeated InterestScalar           interest_scalars   = 9 [(gogoproto.nullable) = false];
+  repeated cosmos.base.v1beta1.Coin utoken_supplies = 10
```
Comment (alexanderbez) [inline proto/umee/leverage/v1beta1/genesis.proto:22]:
```suggestion
  repeated cosmos.base.v1beta1.Coin utoken_supply = 10
```

File: x/leverage/keeper/genesis.go [hunk 2]
```diff
@@ -87,6 +87,7 @@ func (k Keeper) ExportGenesis(ctx sdk.Context) *types.GenesisState {
 		k.GetLastInterestTime(ctx),
 		k.getAllBadDebts(ctx),
 		k.getAllInterestScalars(ctx),
+		k.GetAllUTokenSupplies(ctx),
```
Comment (alexanderbez) [inline x/leverage/keeper/genesis.go:90]:
```suggestion
		k.GetAllUTokenSupply(ctx),
```

**Review by alexanderbez — CHANGES_REQUESTED** [review]

**Review by alexanderbez — APPROVED** [review]
👏

**Comment by toteki** [comment]
Looking into failing e2e

**Pull request merged**